File: Python.PassDetection/src/features/velocity_magnitude_dominant_foot.py
from src.domain.inferences import Feature, Input
from src.domain.recordings import InputData


class VelocityMagnitudeDominantFoot(Feature):

    def calculate(self, input_data: InputData) -> Input:
        dominant_positions = input_data.user_dominant_foot_positions
        timestamps = input_data.timestamps
        values = []

        for i in range(1, len(dominant_positions)):
            dt = timestamps[i] - timestamps[i - 1]
            if dt == 0:
                dt = 1e-6
            dx = (dominant_positions[i].x - dominant_positions[i - 1].x) / dt
            dy = (dominant_positions[i].y - dominant_positions[i - 1].y) / dt
            dz = (dominant_positions[i].z - dominant_positions[i - 1].z) / dt
            magnitude = (dx ** 2 + dy ** 2 + dz ** 2) ** 0.5
            values.append(magnitude)

        values.insert(0, 0.0)  # Insert a zero magnitude at the start
        return Input(self.name, values)

File: Unity.Interactions/Assets/_Project/PassDetection.Common/Features/VelocityMagnitudeDominantFoot.cs
using System.Collections.Generic;
using _Project.PassDetection.Common;
using UnityEngine;

public class VelocityMagnitudeDominantFoot : Feature<InputData>
{
	public override IReadOnlyList<float> Calculate(InputData inputData)
	{
		var dominantPositions = inputData.UserDominantFootPositions;
		var timestamps = inputData.Timestamps;
		var values = new List<float>();

		for (var i = 1; i < dominantPositions.Count; i++)
		{
			var dt = timestamps[i] - timestamps[i - 1];

			if (dt == 0)
				dt = 1e-6f;

			var dx = (dominantPositions[i].x - dominantPositions[i - 1].x) / dt;
			var dy = (dominantPositions[i].y - dominantPositions[i - 1].y) / dt;
			var dz = (dominantPositions[i].z - dominantPositions[i - 1].z) / dt;

			var magnitude = Mathf.Sqrt(dx * dx + dy * dy + dz * dz);
			values.Add(magnitude);
		}

		values.Insert(0, 0.0f); // Insert a zero magnitude at the start
		return values;
	}
}